### ISSUE TITLE
documentation for isa/coerce is incorrect

### ISSUE THREAD

**Issue by icklekitten** [vrurg/MooX-AttributeFilter#4]
It say:

> Filter is called before any other attribute handlers. Its return value is then subject for passing through isa and coerce.

The current implementation actually does coerce and isa *before* filter.

You can test this:

```
use 5.016;
use strict;
use warnings;
use B::Deparse;

{
  package Foo;
  use Moo;
  use MooX::AttributeFilter;
  use Types::Standard qw/Int Num/;
  has xyz => (
    is => "rw",
    isa => Int->plus_coercions(Num, q{ int($_) }),
    filter => sub { "filtered($_[1])" },
    coerce => 1,
  );
}

# says "filtered(100)" showing filter happens *after* type check and coercion
say Foo->new(xyz => 100.001)->xyz;

# shows code of Moo constructor
say B::Deparse->new->coderef2text(Foo->can("new"));
```

This is how I did it in MooseX::AttibuteFilter too.

Should we both change implementation to match documentation, or fix documentation to match implementation?

**Comment by vrurg** [comment]
Basically, the behavior of preceding isa and coerce was unintentional. I only needed the filter to be run at any stage – but not necessarily before anything else. But for now I like the consistency and the idea that isa could be used to catch errors in a complex filter code, for example.

I guess that what happens in your example is different handling for boolean coerce value. Perhaps special measures would be needed to preserve the pattern and more tests to check it.

Thanks!

**Comment by icklekitten** [comment]
The documented behaviour (`filter` before `isa`) makes most sense because if you are going to use output of `filter` to break `isa` why have `isa` at all?

But your current implementation does `isa` before `filter`.

```
use 5.016;
use strict;
use warnings;

{
  package Foo;
  use Moo;
  use MooX::AttributeFilter;
  use Types::Standard qw/Int/;
  has xyz => (
    is => "rw",
    isa => Int,
    filter => sub { "Hello" },
  );
}

my $object = Foo->new(xyz => 1);
say $object->xyz;  # this isn't an Int
```

**Comment by vrurg** [comment]
Not enough testing. That's what it is. The order is inconsistent across different stages: construction calls ISA first, accessor method – next to filter.

I'll see what could be done to this.

**Comment by vrurg** [comment]
I squashed this one in 0.002900. If no major problems would be reported on CPANTS will merge it into stable 0.002002.

**Comment by vrurg** [comment]
Resolved, 0.002002 published.
